captum/optim/models/_common.py

```diff
@@ -254,3 +254,38 @@ class type to replace in the model.
         layers = cast(List[Type[nn.Module]], layers)
         for target_layer in layers:
             replace_layers(model, target_layer, SkipLayer)
+
+
+class MaxPool2dRelaxed(torch.nn.Module):
+    """
+    A relaxed pooling layer, that's useful for calculating attributions of spatial
+    positions. This layer reduces Noise in the gradient through the use of a
+    continuous relaxation of the gradient.
+
+    Args:
+        kernel_size (int or tuple of int): The size of the window to perform max &
+            average pooling with.
+        stride (int or tuple of int, optional): The stride window size to use.
+        padding (int or tuple of int): The amount of zero padding to add to both sides
+            in the nn.MaxPool2d & nn.AvgPool2d modules.
+        ceil_mode (bool, optional): Whether to use ceil or floor for creating the
+            output shape.
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, ...]],
+        stride: Optional[Union[int, Tuple[int, ...]]] = None,
+        padding: Union[int, Tuple[int, ...]] = 0,
+        ceil_mode: bool = False,
+    ) -> None:
+        super().__init__()
+        self.maxpool = torch.nn.MaxPool2d(
+            kernel_size=kernel_size, stride=stride, padding=padding, ceil_mode=ceil_mode
+        )
+        self.avgpool = torch.nn.AvgPool2d(
+            kernel_size=kernel_size, stride=stride, padding=padding, ceil_mode=ceil_mode
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.maxpool(x.detach()) + self.avgpool(x) - self.avgpool(x.detach())
```

Comment:
@NarineK For this layer to work correctly, we want the gradient of the input passed through `nn.AvgPool2d`, while also keep the tensor values from the input passed through `nn.MaxPool2d`. This is what the line does:

```
max_input_grad_avg = self.maxpool(x.detach()) + self.avgpool(x) - self.avgpool(x.detach())
```

Reply:
@ProGamerGov, thank you for the explanation. Does this mean that you want the forward pass to return `self.maxpool(x.detach())` but backward pass to be computed for `self.avgpool(x)` ?
I think it would be great to document it since if we look at it from certain perspective we can say `self.avgpool(x) - self.avgpool(x.detach()) = 0`, so why do we need it.

Reply:
@NarineK Yes, we want the forward pass to return `self.maxpool(x.detach())` while the backward pass uses `self.avgpool(x)`. Lucid uses TensorFlow's gradient override system to perform the same task with slightly different methodology.


The Activation Atlas paper references this [Lucid attribution notebook](https://colab.research.google.com/github/tensorflow/lucid/blob/master/notebooks/building-blocks/AttrSpatial.ipynb#scrollTo=t4_HNz7a9icq) for where the idea was first used. In the notebook, we see the Lucid version of `MaxPool2dRelaxed` has the following description:

> Construct a pooling function where, if we backprop through it, gradients get allocated proportional to the input activation. Then backprop through that instead.

> In some ways, this is kind of spiritually similar to SmoothGrad (Smilkov et al.). To see the connection, note that MaxPooling introduces a pretty arbitrary discontinuity to your gradient; with the right distribution of input noise to the MaxPool op, you'd probably smooth out to this. It seems like this is one of the most natural ways to smooth.

Reply:
Chris helped me a fair bit with recreating the algorithm in PyTorch, so I may not be explaining how it works correctly. I can mark it down that we need to come back to improve the description in a future PR if you want.

Reply:
@ProGamerGov, thank you for the explanation. Do you mind adding this description in the code so that in the future we can understand it easily?

Reply:
@NarineK Okay, I've updated the classes description!